config/initializers/at_exit.rb

```diff
@@ -0,0 +1,61 @@
+#
+# Graceful shutdown of web workers
+#
+# Note this is executed in a different context, Ruby on Rails application is
+# not available, this includes Rails.env, Rails.logger, Rails.root and others.
+#
+# Signal.trap cannot be used as it is redefined in web server (Puma). Multiple
+# at_exit blocks are allowed.
+#
+
+SHUTDOWN_LOGFILE = File.expand_path(__dir__ + "../../../log/shutdown.log")
+PROMETHEUS_DIR = File.expand_path(__dir__ + "../../../tmp/prometheus")
+
+def logmsg(message)
+  File.open(SHUTDOWN_LOGFILE, 'a') { |f| f.puts "[#{Process.pid}] " + message }
+end
+
+def rotatelog
+  return unless File.exist?(SHUTDOWN_LOGFILE)
+  return if File.size(SHUTDOWN_LOGFILE) < 1_000_000
+  FileUtils.rm_f(SHUTDOWN_LOGFILE)
+end
```

Comment:
For production uses we already have `/etc/logrotate.d/foreman`. Would that be sufficient? It may be confusing if a small file is rotated, but part of the logs are missing. Perhaps it should just log to stdout? In production systemd will capture it and in development it may also be easier to see.